macrosynergy/pnl/proxy_pnl_calc.py

```diff
@@ -7,31 +7,174 @@
+from macrosynergy.pnl.transaction_costs import TransactionCosts
+
+
+def get_diff_index(df_wide: pd.DataFrame) -> pd.Index:
+    df_diff = df_wide.diff(axis=0)
+    change_index = df_diff.index[(df_diff.abs() > 0).any(axis=1)]
+    return change_index
+
+
+def _replace_strs(
+    list_of_strs: List[str], old_str: str, new_str: str = ""
+) -> List[str]:
+    return [_.replace(old_str, new_str) for _ in list_of_strs]
+
+
+def _prep_dfs_for_pnl_calcs(
+    df: QuantamentalDataFrame,
+    spos: str,
+    rstring: str,
+) -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame, List[pd.Timestamp]]:
+    df_wide = qdf_to_ticker_df(df=df.copy())
+
+    # Filter
+    u_tickers: List[str] = list(set(df["cid"] + "_" + df["xcat"]))
+    filt_xrs: List[str] = [tx for tx in u_tickers if tx.endswith(rstring)]
+    # do xrs contain `spos` at this point?
+    filt_pos = [tx for tx in u_tickers if tx.endswith(spos)]
+
+    assert set(_replace_strs(filt_xrs, rstring)) == set(
+        _replace_strs(filt_pos, f"_{spos}")
+    )
+
+    # Pivot the dataframes
+    pivot_returns: pd.DataFrame = df_wide.loc[:, filt_xrs]
+    pivot_pos: pd.DataFrame = df_wide.loc[:, filt_pos]
+
+    # warn about NAs
+    dfx: pd.DataFrame
+    for dfx, dfname in [(pivot_returns, "returns"), (pivot_pos, "positions")]:
+        # for each column warns for dates of nas
+        for col in dfx.columns:
+            nas_idx = dfx[col].loc[dfx[col].isna()]
+            if not nas_idx.empty:
+                print(
+                    f"Warning: Series {col} has NAs at the following dates: {nas_idx.index}"
+                )
+
+    # Get the diff index for positions
+    pos_diff_index: pd.DatetimeIndex = get_diff_index(df_wide=pivot_pos)
+    start: str = pivot_pos.first_valid_index()
+    end: str = pivot_pos.last_valid_index()
+
+    # List of rebal_dates - with end date as the last date of the PNL calc
+
+    rebal_dates = sorted(set(pos_diff_index) | {pd.Timestamp(end)})
+
+    # rename cols in pivot_pos and pivot_returns so that they match on mul.
+    pivot_pos.columns = _replace_strs(pivot_pos.columns, f"_{spos}")
+    pivot_returns.columns = _replace_strs(pivot_returns.columns, rstring)
+    pivot_pos = pivot_pos[sorted(pivot_pos.columns)]
+    pivot_returns = pivot_returns[sorted(pivot_returns.columns)]
+
+    return_df_cols = pivot_pos.columns.tolist()
+    pnl_df = pd.DataFrame(index=pd.bdate_range(start, end), columns=return_df_cols)
+
+    return pnl_df, pivot_pos, pivot_returns, rebal_dates
+
+
+def pnl_excl_costs(
+    df: QuantamentalDataFrame,
+    spos: str,
+    rstring: str,
+    pnl_name: str = "PNL",
+) -> pd.DataFrame:
+
+    pnl_df, pivot_pos, pivot_returns, rebal_dates = _prep_dfs_for_pnl_calcs(
+        df=df, spos=spos, rstring=rstring
+    )
+
+    # between each rebalancing date
+    for dt1, dt2 in zip(rebal_dates[:-1], rebal_dates[1:]):
+        curr_pos: pd.Series = pivot_pos.loc[dt1]
+        curr_rets: pd.DataFrame = pivot_returns.loc[dt1:dt2]
```

Comment:
Question: will this include `dt2` in the returns? Goes back to the inclusion of `end` in the rebalance dates. Returns are end-of-day returns, so [1] need to make sure whether they are earned on the day position is taken, or at the end of the period. In either case, for the last period, we should make sure the returns are included for `dt2`.